Migrate wbuchwalter/fairing

wbuchwalter opened the pull request:
Follow up to https://github.com/kubeflow/fairing/pull/2 with correct licence change applied.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/fairing/5)
<!-- Reviewable:end -->

wbuchwalter commented:
/approve

wbuchwalter commented:
@jlewi I will let you merge this one once you have review that licensing is correct.
Note that in this state, fairing does not work with the current version of Kubeflow because of JupyterHub described in https://github.com/kubeflow/fairing/issues/6.

I am working on a fix. I will open a PR here once it is merged.

Also not that @ashahba has been working on a PR to add a new builder using configmaps instead of Knative. This PR will be transferred here as well so it can be discussed by the community.

jlewi commented:
/assign @r2d4 

@r2d4 Can you take a lead on reviewing this please?

r2d4 commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "python.pythonPath": "/home/will/envs/DoWhy-zTd0ka9x/bin/python"
```
I think this file can be removed

r2d4 commented on Dockerfile.dev:
```diff
@@ -0,0 +1,11 @@
+RUN pip install -e .
+
+COPY ./ /opt/fairing
+RUN python ./setup.py install
```
nit: add a newline

r2d4 commented on Dockerfile.jupyter:
```diff
@@ -0,0 +1,11 @@
+RUN pip install -e .
+
+COPY ./ /opt/fairing
+RUN python ./setup.py install
```
nit: add a newline

r2d4 commented on Makefile:
```diff
@@ -0,0 +1,22 @@
+		echo "FAIRING_DEV_DOCKER_USERNAME is unset. Set it to your docker username to use this command"; \
+	else \
+		pip install . ;\
+		docker build -t ${FAIRING_DEV_DOCKER_USERNAME}/fairing:latest -f Dockerfile.dev . ; \
```
I have some ideas on how we could get rid of scripts like this, but for now you might want to avoid pushing anything with "latest" and tag with git commit sha.

wbuchwalter replied:
This specific script is to make easier to help developers try their changes without having to go through the pip release process.
You would set `FAIRING_DEV_DOCKER_USERNAME` to your username and this would push a dev version to your dockerhub. This image is then pulled instead of the regular base image.
So latest is needed so that you don't have to go modify this script on every change.

r2d4 commented on Dockerfile.jupyter:
```diff
@@ -0,0 +1,11 @@
+FROM tensorflow/tensorflow:latest-py3
```
Let pin to a specific version of tf?

r2d4 commented on README.md:
Maybe this README should be moved somewhere within kubeflow/*?

r2d4 commented on README.md:
```diff
@@ -1,2 +1,230 @@
+* `package`: Defines the repository (this could be your DockerHub username, or something like `somerepo.acr.io` on Azure for example) and name that should be used to build the image. You can control wether you want to publish the image by setting `publish` to `True`.
+* `strategy`: Specify which training strategy should be used (more details below).
+* `architecture`: Specify which architecture should be used. (more details below)
+* `tensorboard`: [Optional] If specified, will spawn an instance of TensorBoard to monitor your trainings
```
Should this be the responsibility of `fairing`? (Not a leading question, I'm still understanding the whole project).

r2d4 commented on README.md:
Shouldn't distributed training be the default? Why would you not want to run it distributed if you're running it on kubeflow?

r2d4 commented on README.md:
Can you explain why the PVC is needed?

r2d4 replied:
I was thinking that fairing would be responsible for converting the notebook to a container, and then spawning the correct TF/Pytorch job through the TFJob CR.

r2d4 commented on examples/distributed-training/main.py:
```diff
@@ -0,0 +1,248 @@
+
+if __name__ == '__main__':
+    model = MyModel()
+    model()
```
nit: newline

r2d4 commented on examples/distributed-training/requirements.txt:
```diff
@@ -0,0 +1 @@
+tensorflow
```
pin this to a version

r2d4 commented on examples/hyperparameter-tuning/requirements.txt:
```diff
@@ -0,0 +1 @@
+tensorflow
```
pin this to a version

r2d4 commented on examples/kubeflow-jupyter-notebook/README.md:
```diff
@@ -0,0 +1,66 @@
+See: [Installing Knative Build Component](https://github.com/knative/docs/blob/master/build/installing-build-component.md)
+
+
+### Container Registry Secret
```
I think that we should reuse the user service account that gets created instead

wbuchwalter replied:
I'm not sure I understand with service account you are referring to?

r2d4 commented on examples/kubeflow-jupyter-notebook/README.md:
```diff
@@ -0,0 +1,66 @@
+
+Kubeflow needs to be pre-installed in the cluster.
+
+### Knative Build Component
```
I think that Knative is a heavy dependency to build images on the cluster. If we're just appending python files, we can do it much simpler and quicker without knative.

wbuchwalter replied:
Knative is the solution that came out after discussing with @jlewi.
You can see some of the context for that in this thread: https://github.com/kubeflow/kubeflow/issues/1240#issuecomment-408878266

Note that @ashahba has a PR opened on the old repo to use configmaps directly and bypass the build step entirely (https://github.com/wbuchwalter/fairing/pull/27). I ll move this PR here once this one get merged. However this approach does not allow for a lot of flexibility since you cannot customize the remote environment.

Curious to hear your idea.

r2d4 commented on fairing/architectures/kubeflow/distributed.py:
```diff
@@ -0,0 +1,49 @@
+        self.ps_count = ps_count
+        self.worker_count = worker_count
+
+    def add_jobs(self, svc, count, img, name, volumes, volume_mounts):
```
WDYT about using the k8s clients directly here?

r2d4 commented on fairing/backend/native.py:
```diff
@@ -0,0 +1,76 @@
+        return svc
+    
+    def stream_logs(self, image_name, image_tag):
+        mp_client = MetaparticleClient()
```
Why the dependency on metaparticle? It seems a bit unnecessary

wbuchwalter replied:
Context for metaparticle here: https://github.com/kubeflow/kubeflow/issues/1240#issuecomment-408493465.
However I am not convinced anymore that this is the right approach since metaparticle is not maintained it seems.
We can remove it in a later PR.

r2d4 commented on fairing/builders/knative/knative.py:
```diff
@@ -0,0 +1,103 @@
+                                             default='/src/Dockerfile')]
+        volume_mount = kubernetes.client.V1VolumeMount(name='src', mount_path='/src')
+        steps = [BuildTemplateSpecStep(name='build-and-push',
+                                       image='gcr.io/kaniko-project/executor',
```
nit: pin to a version.

If we're hardcoding kaniko here, it might be simple to just launch the kaniko job directly instead of calling it through knative. But it might be easier to skip kaniko and build CRD altogether.

wbuchwalter commented:
@r2d4 Addressed some of your comments.
For everything related to removing metaparticle and using k8s client directly instead, I agree with the direction however I would like that to be done in a follow-up PR so that I don't block all the progress of other PRs.
Same applies to Knative, I am not specifically attached to Knative and I am curious to hear what you propose, but it would be nicer to discuss this in a separate issue and implement what is decided in a follow-up PR.

What do you think?

r2d4 commented:
Yep SGTM.

/approve
/lgtm
